plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -417,11 +417,92 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
          packed_transaction_ptr  trx;
          next_func_t             next;
       };
+   
       // The queue storing read-only transactions to be executed by read-only threads
-      struct ro_trx_queue_t {
-         std::mutex              mtx;
-         std::deque<ro_trx_t>    queue;
+      class ro_trx_queue_t {
+      public:
+         void push_back(ro_trx_t&& trx) {
+            std::unique_lock<std::mutex> g( mtx );
+            queue.push_back(std::move(trx));
+            notify_waiting(g, false);
+         }
+         
+         void push_front(ro_trx_t&& trx) {
+            std::unique_lock<std::mutex> g( mtx );
+            queue.push_front(std::move(trx));
+            notify_waiting(g, false);
+         }
+
+         bool empty() const {
+            std::lock_guard<std::mutex> g( mtx );
+            return queue.empty();
+         }
+
+         // may wait if the queue is empty, and not all other threads are already waiting.
+         // returns true if a transaction was dequeued and should be executed, or false 
+         // if conditions are met to stop processing transactions. 
+         bool pop_front(ro_trx_t &trx) {
+            std::unique_lock<std::mutex> g( mtx );
+            
+            if (should_exit()) {
+               notify_waiting(g, true);
+               return false;
+            }
+
+            if (queue.empty()) {
+               if (num_waiting + 1 == max_waiting) {
+                  notify_waiting(g, true);
+                  return false;
+               }
+               
+               ++num_waiting;
+               cond.wait(g);
+               --num_waiting;
```

Comment:
This is a very odd use of a condition variable. I would recommend reviewing: https://embeddedartistry.com/blog/2022/01/10/remember-to-lock-around-all-stdcondition_variable-variables/

Reply:
Can you elaborate on what is odd?
As expected, the wait() is called when the mutex lock is active, and the `notify_*` is done after unlocking.

Reply:
OK, got it, spurious notifications were not taken into account. Current version should be safe now.